feat: improve bounds checking

rrsettgast opened the pull request:
This PR adds further bounds checking in GEOS.

- [x] std::vector wrapper that has bounds checking for `::operator[]( index )`
~~- [ ] `std::map/unordered_map` wrapper that has bounds checking for `::operator[]( index )`~~
~~- [ ] disallow use of `map/unordered_map::operator[]` to perform insertion.~~

the new wrappers are preprocessor guarded with aliases s.t. they alias to the std::container in the non-bounds checking case.

rrsettgast committed:
add wrapper for std::vector and std::map and std::unordered_map

rrsettgast committed:
rename some cmake variables

rrsettgast committed:
pre-merge commit...just some renaming

rrsettgast committed:
Merge branch 'develop' into feature/improveBoundsChecking

rrsettgast committed:
make a std::vector wrapper so we can turn on bounds checking

rrsettgast committed:
Merge branch 'develop' into feature/improveBoundsChecking

rrsettgast committed:
more substitution

rrsettgast commented:
@OmarDuran I think this would solve most issues with bounds checking of `std::vector`

OmarDuran commented:
> @OmarDuran I think this would solve most issues with bounds checking of `std::vector`

@rrsettgast  Thanks for fixing this, hopefully the CI will be testing bounds soon. I noticed another issue in [VTKUtilities](https://github.com/GEOS-DEV/GEOS/blob/4aa2c9fd542b18cbbb493580a461f53cb33f2e42/src/coreComponents/mesh/generators/VTKUtilities.cpp#L1438) where nodeTri0 has size 3. This clearly fails in Debug mode, so it’ll need to be addressed as well.

rrsettgast committed:
turn bounds checks on for CI jobs...add a non-bounds check job for CPU runs

rrsettgast committed:
some uncrustify...and guarded LvArray test from incorrect gcc warning

rrsettgast committed:
modify guards

rrsettgast committed:
modify guards

rrsettgast committed:
disable bounds check on gcc11 due to invalid warning

rrsettgast committed:
fix ci option for bounds check

arng40 commented:
@rrsettgast What would you like us to do?

rrsettgast committed:
change NodeManager::setGeometricalRelations calls to assimilate s.t. they assume unsorted with duplicates

rrsettgast committed:
Merge branch 'develop' into feature/improveBoundsChecking

wrtobin reviewed (approved):
looks good.
- do we want to remove the commented-out map wrappers? 
- im not the biggest fan of `stdVector` as the name, but don't really have a preferred alternative, so if we wanna go with that for now its fine

rrsettgast committed:
documentation for stdVectorWrapper

rrsettgast commented:
@MelReyCG @arng40 

Can you guys look into the 2nd and third task:

- [ ] std::map/unordered_map wrapper that has bounds checking for ::operator[]( index )
- [ ] disallow use of map/unordered_map::operator[] to perform insertion.

rrsettgast committed:
submodule update

MelReyCG commented on src/coreComponents/common/StdContainerWrappers.hpp:
```diff
@@ -0,0 +1,161 @@
+  /// Type alias for the base class (i.e., std::vector)
+  using Base = std::vector< T, Allocator >;
+
+  /// inherit constructors from the base class
+  using Base::Base;
+
+  /// inherit std::vector::push_back()
+  using Base::push_back;
+
+  /// inherit std::vector::size()
+  using Base::size;
+
+  /// inherit std::vector::empty()
+  using Base::empty;
```
Why do these usings are needed? As you are inheriting publicly from `std::vector`, aren't they already visible?

rrsettgast replied:
yeah...they certainly should be...I don't recall why I put these in....perhaps copilot just started to do it and I didn't think about it.

rrsettgast replied:
I removed them.

MelReyCG commented:
The started work here really is a good thing, `GEOS_ENABLE_BOUNDS_CHECK` not having effect on `std` container was problematic.
We can imagine adding some exceptions with more information in the future, in case of out-of-bound or not-existing key.

MelReyCG commented:
@arng40 and / or me will work on those 2 last tasks. After that, we will also test `GEOS_ENABLE_BOUNDS_CHECK` on the last crashing cases we got.

rrsettgast committed:
remove unrequired declarations in stdVector

rrsettgast committed:
Merge branch 'develop' into feature/improveBoundsChecking

rrsettgast committed:
Update ci_tests.yml

disable range checking on cuda builds...for some reason they are stalling in this PR??? strange
